bpo-38377: multiprocessing.SemLock requires working /dev/shm on Linux.

mcepl opened the pull request:
Skip test_get_event_loop_new_process on systems which don’t provide it
(e.g., building environments of some Linux distributions).

<!-- issue-number: [bpo-38377](https://bugs.python.org/issue38377) -->
https://bugs.python.org/issue38377
<!-- /issue-number -->


Automerge-Triggered-By: @pitrou

mcepl committed:
_multiprocessing.SemLock requires working /dev/shm on Linux.

Skip test_get_event_loop_new_process on systems which don’t provide it
(e.g., building environments of some Linux distributions).

vstinner commented on Lib/test/test_asyncio/test_events.py:
```diff
@@ -2658,7 +2659,8 @@ def tearDown(self):
     if sys.platform != 'win32':
-
+        @unittest.skipUnless(os.path.exists("/dev/shm") and stat.S_IMODE(os.stat('/dev/shm').st_mode) >= 0o777,
+                             "Doesn't work without /dev/shm.")
```
I'm not sure if /dev/shm should be hardcoded here. This check looks to be specific to Linux, no?

Maybe use Arena._dir_candidates of multiprocessing.heap. I'm not sure what to do if _dir_candidates is empty (platforms other than Linux).

I dislike that test_asyncio becomes tidely coupled to multiprocessing internals. Is there a function in multiprocessing somewhere to check if /dev/shm is required and usable?

pitrou replied:
Instead of that, why not skip if there is an environment variabled named e.g. `PYTHON_NO_DEV_SHM`?

mcepl replied:
@pitrou I would still hoped for some kind of magic, what @vstinner suggested. Any ideas?

pitrou replied:
I think it's more flexible to let you handle the magic in your build configuration. I don't think we want to handle all weird corner cases in our test suite.

mcepl replied:
OK, so the patch with ``unittest.skipIf('PYTHON_NO_DEV_SHM' in os.environ)`` would be accepted?

pitrou replied:
IMHO it should. I'll let @vstinner comment.
(or perhaps `os.environ.get('PYTHON_NO_DEV_SHM') == '1'`)

vstinner replied:
I don't understand the relationship between asyncio, multiprocessing and /dev/shm.

mcepl committed:
Rewrite according to the review on the upstream pull request.

Instead of doing things automagically, add new environmental
variable PYTHON_NO_DEV_SHM, which switches off particular tests.

eamanu commented on Misc/NEWS.d/next/Tests/2019-10-05-15-50-24.bpo-38377.gKfVDt.rst:
```diff
@@ -0,0 +1,5 @@
+Add the environment variable :data:`PYTHON_NO_DEV_SHM`,
+which when set to 1, makes the Python skip tests depending
+on :data:`/dev/shm` interface, which are missing in building
+environments of some Linux distributions. Currently skipping
```
IMHO we cannot support something that is not used for linux "standard"?

I am not an expert, but this sound like a specific distribution issue, and will be a little difficult (maybe)
pay attention for all Linux distributions issues.

mcepl replied:
I don’t understand what’s your point. Do you think we should care less about Linux distributions, or what?

eamanu replied:
Hmm I think so, I saw some other projects that when the issues come from an specific distribution, that can be fixed simply with a patch on the distribution package, and don't try to skip (in this case the test) on the upstream. It's just an opinion :)

mcepl replied:
OK, this is just an opinion and not an action item.

pitrou reviewed (approved):
+1 from me

mcepl committed:
Use correct Sphinx role for the Python method.

tiran commented on Lib/test/test_asyncio/test_events.py:
```diff
@@ -2671,7 +2671,8 @@ def tearDown(self):
 
     if sys.platform != 'win32':
-
+        @unittest.skipIf(os.environ.get('PYTHON_NO_DEV_SHM') == '1',
```
nit-pick, you should also check ``sys.flags.ignore_environment`` here.
